scripts/build.sh

```diff
@@ -5,63 +5,109 @@
 ALL_XC_ARCH="386 amd64 arm arm64 ppc64le"
 ALL_XC_OS="linux darwin windows freebsd openbsd solaris"
 
+# Exit immediately if a command fails
 set -e
 
+# Validates that a necessary tool is on the PATH
+function validateToolPresence
+{
+    local TOOLNAME=$1
+    which ${TOOLNAME} >/dev/null || echo "${TOOLNAME} is not on the path. Exiting..."
+    exit 1
+}
+
+# Validates that all used tools are present; exits when any is not found
+function validatePreconditions
+{
+    echo "==> Checking for necessary tools..."
+    validateToolPresence realpath
+    validateToolPresence dirname
+    validateToolPresence tr
+    validateToolPresence find
+}
+
 # Get the parent directory of where this script is.
-SOURCE="${BASH_SOURCE[0]}"
-while [ -h "$SOURCE" ] ; do SOURCE="$(readlink "$SOURCE")"; done
-DIR="$( cd -P "$( dirname "$SOURCE" )/.." && pwd )"
-
-# Change into that directory
-cd $DIR
-
-# Delete the old dir
-echo "==> Removing old directory..."
-rm -f bin/*
-rm -rf pkg/*
-mkdir -p bin/
-
-# helpers for Cygwin-hosted builds
-: ${OSTYPE:=`uname`}
-
-case $OSTYPE in
-    MINGW*|MSYS*|cygwin|CYGWIN*)
-        # cygwin only translates ';' to ':' on select environment variables
-        PATHSEP=';'
-        ;;
-    *)	PATHSEP=':'
-esac
-
-function convert_path() {
-    local flag
-    [ "${1:0:1}" = '-' ] && { flag="$1"; shift; }
+# NOTE: I'm unsure why you don't just use realpath like below
+function enterPackerSourceDir
+{
+    echo "==> Entering Packer source dir..."
+    local BUILD_SCRIPT_PATH="${BASH_SOURCE[0]}"
+    SOURCEDIR=$(dirname $(dirname $(realpath "${BUILD_SCRIPT_PATH}")))
+    cd ${SOURCEDIR}
+}
+
+function ensureOutputStructure {
+    echo "==> Ensuring output directories are present..."
+    mkdir -p bin/
+    mkdir -p pkg/
+}
+
+function cleanOutputDirs {
+    echo "==> Removing old builds..."
+    rm -f bin/*
+    rm -fr pkg/*
+}
 
-    [ -n "$1" ] || return 0
-    case ${OSTYPE:-`uname`} in
-        cygwin|CYGWIN*)
-            cygpath $flag -- "$1"
+function lowerCaseOSType {
+    local OS_TYPE=${OSTYPE:=`uname`}
+    echo "${OS_TYPE}" | tr "[:upper:]" "[:lower:]"
+}
+
+# Returns the OS appropriate path separator
+function getPathSeparator {
+    # helpers for Cygwin-hosted builds
+    case "$(lowerCaseOSType)" in
+        mingw*|msys*|cygwin*)
+            # cygwin only translates ';' to ':' on select environment variables
+            echo ';'
             ;;
-        *)  echo "$1"
+        *)	echo ':'
     esac
 }
 
+function convertPathOnCygwin() {
+    local flag
+    local somePath
+    if [ "${1:0:1}" = '-' ]; then
```

Comment:
The "optional" flag could be relegated to second parameter which would make logical flow better, but then it wouldn't look command-like (if that's the intent)